File: 백준/Gold/27172. 수 나누기 게임/수 나누기 게임.py
import sys;input = sys.stdin.readline
n = int(input())
arr = list(map(int, input().split()))
ans = [0] * 1000001
_set = set(arr)
_max = max(arr)
for i in arr:
    tem = i * 2
    while tem <= _max:
        if tem in _set:
            ans[i] += 1
            ans[tem] -= 1
        tem += i
for i in arr:
    print(ans[i], end = ' ')

File: 백준/Gold/27172. 수 나누기 게임/수 나누기 게임.cc
#include<iostream>
#include<algorithm>
#include<unordered_set>
using namespace std;
int n;
int arr[100001];
int ans[1000001];
int tem;
int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cin >> n;
	for (int i = 0; i < n; i++) {
		cin >> arr[i];
	}
	int *_max = max_element(begin(arr), end(arr));
	unordered_set<int> set(arr, arr + sizeof(arr)/sizeof(arr[0]));
	for (int i=0; i<n; i++){
		int x = arr[i];
		tem = x << 1;
		while (tem <= *_max) {
			if (set.find(tem) != set.end()) {
				ans[x]++;
				ans[tem]--;
			}
			tem += x;
		}
	}
	for (int i = 0; i < n; i++) {
		cout << ans[arr[i]] << " ";
	}
}